internal/backend/backend.go

```diff
@@ -167,13 +179,41 @@ func (b *Backend) Close(ctx context.Context) error {
 	return nil
 }
 
-func (b *Backend) getUserID(username string, password []byte) (string, error) {
+func (b *Backend) getUserID(ctx context.Context, username string, password []byte) (string, error) {
+	if atomic.LoadInt32(&b.loginErrorCount) >= maxLoginAttempts {
+		b.lockLogin.Lock()
+		atomic.StoreInt32(&b.loginErrorCount, 0)
+		b.lockLogin.Unlock()
+	}
+
 	for _, user := range b.users {
 		if user.connector.Authorize(username, password) {
+			atomic.StoreInt32(&b.loginErrorCount, 0)
 			return user.userID, nil
 		}
 	}
 
+	atomic.AddInt32(&b.loginErrorCount, 1)
+
+	if atomic.LoadInt32(&b.loginErrorCount) == maxLoginAttempts {
+		b.notifyLoginBlocked(true)
```

Comment:
The connector should internally handle this. Keep in mind that one connector could potentially authorize a lot of different users. In this case you are blocking the login for all the users rather than the one where the authentication actually failed.

Reply:
On login fails we don't know what is the user that failed to login, do we ?
Idea here was to notify any connector that we locked the Login procedure for a while and then the Connector can decide what we want to do with that.
For simple implementation I was just thinking of printing text at login page. Rather than a toggle off mechanism that is a bit more advanced.

Reply:
but still doable (toggle off mechanism from connector)

Reply:
I understand the connector/user oriented part if the login attempt  target a known user, but how can we handle login attempts that are not targeting known user ?

I also understand that by doing global tempo/lock, we can easily deny the Login mechanism for all connector by spamming wrong Login.
Also only handling known user could provide a hint while trying to force the login since we can identify by the locking mechanism that the user is known. (but at least still better than denying the login process).

If we go that way, there is nothing to do within Gluon and everything should be handle on connector side with the Authorize method (that is still called for every connector until one says OK).
Is that what you have in mind ?

Reply:
I was trying to say we could keep the current logic but without the extra connector method.

 If we really wanted to implement more advanced error pattern matching we need to resort to IP filtering. An interface to implement the behavior would be the better option. That way once could tailor the behavior for Bridge and any other users of the library.

The block all equivalent could be exactly the code you have submitted here.